virtcontainers/qemu.go

```diff
@@ -1275,6 +1275,69 @@ func (q *qemu) hotplugVhostUserDevice(vAttr *config.VhostUserDeviceAttrs, op ope
 	return nil
 }
 
+// Query QMP to find the PCI slot of a device, given its QOM path or ID
+func (q *qemu) qomGetSlot(qomPath string) (uint8, error) {
+	addr, err := q.qmpMonitorCh.qmp.ExecQomGet(q.qmpMonitorCh.ctx, qomPath, "addr")
+	if err != nil {
+		return 0xff, err
+	}
+	addrf, ok := addr.(float64)
+	// XXX going via float makes no real sense, but that's how
+	// JSON works, and we'll get away with it for the small values
+	// we have here
+	if !ok {
+		return 0xff, fmt.Errorf("addr QOM property of %s is %t not a number", qomPath, addr)
+	}
+	addri := uint8(addrf)
+
+	if (addri & 0x7) != 0 {
+		return 0xff, fmt.Errorf("Unexpected non-zero PCI function on %s", qomPath)
+	}
+
+	return addri >> 3, nil
+}
+
+// Query QMP to find a device's PCI path given its QOM path or ID
+// "PCI path" means the PCI slot number of each bridge leading to the
+// device, and finally the device itself, as a '/' separated list
+func (q *qemu) qomGetPciPath(qemuID string) (string, error) {
+	// XXX: For now we assume there's exactly one bridge, since
```

Comment:
what about root ports ? can this implementation support more than one root port?

Reply:
For the purposes here, a root port's virtual PCI to PCI bridge is basically the same as an actual PCI to PCI bridge.  Either way we have one bridge (either a root port or PCIE->PCI bridge) and then the device.
It can certainly handle multiple root ports - which root port is encoded into that bridge slot.

What it can't handle is a bridge under a bridge, including a bridge under a root port.  That also includes PCI-E switches, which incorporate 2 stacked virtual P2P bridges within them.  But we don't allow any of those configurations for the time being.

I hope to generalize the handling of PCI paths at some point, but I'm hoping that can wait until Kata2.